cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,166 @@
+ * matching rows between `left_on` and `right_on`. The first column corresponds to
+ * rows in `left_on`, and the second to `right_on`. A value of `-1` in the second column
+ * indicates that the corresponding row in `left_on` has no match. It is the caller's
+ * responsibility to free these columns.
+ * @param[in] join_context The context to use to control how
+ * the join is performed,e.g., sort vs hash based implementation
+ * 
+ * @returns Result of joining `left` and `right` tables on the columns
+ * specified by `left_on` and `right_on`. The resulting table will be joined columns of
+ * `left` and then joined columns of `right`. For any "common" columns indicated
+ * by `columns_in_common`, only a single output column is produced and the order of the
+ * columns will be same as `left` and then `right`.
+ */
+cudf::table left_join(
+                         cudf::table const& left,
+                         cudf::table const& right,
+                         std::vector<gdf_size_type> const& left_on,
+                         std::vector<gdf_size_type> const& right_on,
+                         std::vector<std::pair<gdf_size_type, gdf_size_type>> const& columns_in_common,
+                         cudf::table *joined_indices,
+                         gdf_context *join_context);
+
+/** 
+ * @brief  Performs a full join (also known as full outer join) on the
+ * specified columns of two tables (left, right)
+ * 
+ * @example TableA a:{0,1,2}
+ *          TableB b:{1,2,3}, a:{1,2,5} And column 'a' of TableB joins
+ *          column 'a' of TableA.
+ * The result would be Table-res a:{0, 1, 2, 5}, b:{NULL, 1, 2, 3}
+ * Result is union of column 'a' from Table A and column 'a' from Table B
+ *
+ * @throws cudf::logic_error
+ * if a sort-based join is requested and either `right_on` or `left_on` contains null values.
+ *
+ * @param[in] left The left table
+ * @param[in] right The right table
+ * @param[in] left_on The column's indices from `left` to join on.
+ * Column `i` from `left_on` will be compared against column `i` of `right_on`.
+ * @param[in] right_on The column's indices from `right` to join on.
+ * Column `i` from `right_on` will be compared with column `i` of `left_on`.
+ * @param[in] columns_in_common is a vector of pairs of column indices into
+ * `left_on` and `right_on`, respectively, that are "in common". For "common"
+ * columns, only a single output column will be produced, which is gathered
+ * from respective columns in `left_on` and `right_on`. Else, for every column in
+ * `left_on` and `right_on`, an output column will be produced.
+ *
+ * @param[out] * @returns joined_indices Optional, if not `nullptr`, on return, will
+ * contain two non-nullable, `GDF_INT32` columns containing the indices of
+ * matching rows between `left_on` and `right_on`. The first column corresponds to
+ * rows in `left_on`, and the second to `right_on`. A value of `-1` in the second column
```

Comment:
For a full join, -1 values can appear in the indices corresponding to both the left and right table, indicating non-matches.